### ISSUE TITLE
FP redundant_clone \w  unsized_locals unsized_fn_params

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9953]
### Summary

.

### Lint Name

redundant_clone

### Reproducer

I tried this code:

```rust
#![allow(incomplete_features)]
#![feature(unsized_locals, unsized_fn_params)]

pub trait Foo {
    fn foo(self) -> String;
}

impl Foo for str {
    fn foo(self) -> String {
        self.to_owned()
    }
}

fn main() {}
```

I saw this happen:

```
after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0308]: mismatched types
  --> src/main.rs:10:9
   |
9  |     fn foo(self) -> String {
   |                     ------ expected `std::string::String` because of return type
10 |         self
   |         ^^^^- help: try using a conversion method: `.to_string()`
   |         |
   |         expected struct `std::string::String`, found `str`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.

warning: redundant clone
  --> src/main.rs:10:13
   |
10 |         self.to_owned()
   |             ^^^^^^^^^^^ help: remove this
   |
note: this value is dropped without further use
  --> src/main.rs:10:9
   |
10 |         self.to_owned()
   |         ^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_clone
   = note: `#[warn(clippy::redundant_clone)]` on by default
```

### Version

```text
rustc 1.67.0-nightly (b3bc6bf31 2022-11-24)
binary: rustc
commit-hash: b3bc6bf31265ac10946a0832092dbcedf9b26805
commit-date: 2022-11-24
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```
